### PR TITLE
Issue/angular ssr without serverless

### PR DESCRIPTION
**Pull request by niekraaijmakers**
## Description

This PR adds an option to generate a module for angular, that will make it very easy to set up server-side rendering in with express for the SPA editor.

## Related Issue

https://github.com/adobe/aem-project-archetype/issues/724

## Motivation and Context

Our partners and customers are having a hard time learning / understanding and setting up the initial work for server side rendering in AdobeIO runtime or AWS lambda. Server side rendering is by nature a very complex topic. This will make getting started much easier, as it provides a point of "hello world" which works in express.

Unfortunately, there is no clean solution I figured out to achieve SSR using IO runtime with Angular.
I found a way to make it work but it involved too much hacking to be in the archetype.
The branch will remain on my fork.
If the libraries the solution depends on get's overhauled or another solution is presented, we can consider it.

## How Has This Been Tested?

Aside from the automatic tests, I deployed the projects from the generated target folder to AEM.
I then ran express, and checked that the server-side content came up and checked if the browser was fine.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/main/archetype/ui.frontend.angular/README.md [hunk 1]
```diff
@@ -40,3 +40,134 @@ Furthermore, an AEM ClientLib is generated from the app using the [`aem-clientli
...
+
+
+
+#if ( $enableSSR == "y" )
+
+${hash} Frontend Serverless Build (Adobe IO Runtime)
+
+${hash}${hash} Features
+
+* AdobeIO Runtime (Serverless) rendering of the SPA editor
```
Comment (CezCz) [inline src/main/archetype/ui.frontend.angular/README.md:52]: is this true? is it not only express here?

File: src/main/archetype/ui.frontend.angular/README.md [hunk 2]
```diff
@@ -40,3 +40,134 @@ Furthermore, an AEM ClientLib is generated from the app using the [`aem-clientli
...
+
+Finally, the initialization script of the app is different:
+
+main.ts:
+```js
+const initialStateScriptTag = document.getElementById('__AEM_STATE__');
+if(!!initialStateScriptTag) {
+  try {
+    const initialState = JSON.parse(initialStateScriptTag.innerHTML);
+    // @ts-ignore
```
Comment (CezCz) [inline src/main/archetype/ui.frontend.angular/README.md:137]: I guess it's fine to remove @ts-ignores from examples ?

File: src/main/archetype/ui.frontend.angular/README.md [hunk 3]
```diff
@@ -40,3 +40,134 @@ Furthermore, an AEM ClientLib is generated from the app using the [`aem-clientli
...
+  }catch(err){
+    console.warn('failed to hydrate app', err);
+  }
+}
+```
+
+app.component.ts:
+```js
+
+if(isPlatformBrowser(_platformId)){
```
Comment (CezCz) [inline src/main/archetype/ui.frontend.angular/README.md:149]: where is `_platformId` coming from? Maybe we want to show the import and in general what this variable represents

File: src/main/archetype/ui.frontend.angular/src/app/app.component.ts [hunk 4]
```diff
@@ -27,11 +35,27 @@ export class AppComponent {
   items: any;
   itemsOrder: any;
   path: any;
-
+#if ( $enableSSR == "n")
```
Comment (CezCz) [inline src/main/archetype/ui.frontend.angular/src/app/app.component.ts:38]: can `$enableSSR` have only two values? "n" and "y" ? If not (or not sure) then maybe one of the ifs (probably "n") should be used as fallback?

File: src/main/archetype/ui.frontend.angular/src/main.ts [hunk 5]
```diff
@@ -29,6 +29,19 @@ if (environment.production) {
   }
 }
 
+const initialStateScriptTag = document.getElementById('__AEM_STATE__');
+if(!!initialStateScriptTag) {
```
Comment (CezCz) [inline src/main/archetype/ui.frontend.angular/src/main.ts:33]: it's fine to not convert this to a boolean.

**Review by CezCz — APPROVED** [review]
lgtm, please address my comments how you like and proceed :)

File: src/main/archetype/ui.frontend.angular/README.md (hunk 1 above)
Comment (niekraaijmakers) [reply src/main/archetype/ui.frontend.angular/README.md:52]: Ah yep, good catch

File: src/main/archetype/ui.frontend.angular/README.md (hunk 2 above)
Comment (niekraaijmakers) [reply src/main/archetype/ui.frontend.angular/README.md:137]: nope, it won't compile :(

File: src/main/archetype/ui.frontend.angular/src/app/app.component.ts (hunk 4 above)
Comment (niekraaijmakers) [reply src/main/archetype/ui.frontend.angular/src/app/app.component.ts:38]: yep only these 2 values

File: src/main/archetype/ui.frontend.angular/README.md [hunk 6]
```diff
@@ -40,3 +40,134 @@ Furthermore, an AEM ClientLib is generated from the app using the [`aem-clientli
...
+   * docker pull adobeapiplatform/adobe-action-nodejs-v10:3.0.21
+
+${hash}${hash} Usage
+
+The following npm scripts are available:
+
+* `npm run build-serverless` - Build the serverless code
+* `npm run build-server` - Build express code (debuggable with sourcemaps)
+* `npm run start-ssr-express` - Runs the express code (debuggable with sourcemaps)
+* `npm run start-ssr-ioruntime` - Deploys the serverless code to a local docker container with AIO CLI. Requires docker running. Enables you to debug with visual studio code.
```
Comment (znikolovski) [inline src/main/archetype/ui.frontend.angular/README.md:71]: Based on the feedback above I take it the runtime run scripts also don't apply?

Comment (niekraaijmakers) [reply src/main/archetype/ui.frontend.angular/README.md]: updated

**Pull request merged**